{% extends 'base.html' %}
{% load staticfiles %}

{% block content %}
<div class="header">
    <div class="container">
        <div id="menu">
            <ul id="menubar">
                <li><a href="{% url 'index' %}">HOME</a></li>
                {% if user %}
                    <li><a href="{% url 'pipeline' %}">PIPELINES</a></li>
                {% else %}
                    <li><a href="{% url 'pipeline' %}" class="disabled"><span style="text-decoration: line-through;">PIPELINES</span></a></li>
                {% endif %}
                <li><a href="{% url 'delete' %}">DELETE</a></li>
                <li><a href="{% url 'readme' %}">README</a></li>
                <div id="smenubar">
                    {% if user %}
                        <li>
                            <b><small style="color:rgb(136,162,175); font-size:9pt;">Logged in as: <span style="color:rgb(255,100,0);">{{ user }}</span></small></b>
                        </li>
                        <li><a href="{% url 'logout' %}">LOGOUT</a></li>
                            {% if super %}
                                <li>&nbsp;</li><li>&nbsp;</li><li>&nbsp;</li><li>&nbsp;</li><li>&nbsp;</li><li>&nbsp;</li>
                                <li>&nbsp;</li><li>&nbsp;</li><li>&nbsp;</li><li>&nbsp;</li><li>&nbsp;</li><li>&nbsp;</li>
                                <li>&nbsp;</li><li>&nbsp;</li><li>&nbsp;</li><li>&nbsp;</li><li>&nbsp;</li>
                                <li><a href="{% url 'signup' %}">SIGN UP</a></li>
                            {% endif %}
                    {% else %}
                        <li><a href="{% url 'login' %}">LOGIN</a></li>
                    {% endif %}
            </ul>
        </div>
    </div>
</div>
<div class="container" style="margin-top:100px">
    <p>
        <h2>Pipeline Tools</h2>
        <ul>
            <li>Flye 2.4.2</li>
            <li>Miniamp 2.15-r915</li>
            <li>Miniasm 0.3-r179</li>
            <li>resFinder 2.1</li>
            <li>BLAST+ 2.8.1</li>
        </ul>
    </p>
    <div id="read_analysis">
        <h2>Nanopore Analysis Pipeline</h2>
        <h3>Basecalling, de-barcoding and FASTQ conversion</h3>
        <p>
            Albacore is used to read the raw FAST5 files and will do another run of basecalling with de-barcoding <br>
            and will convert the reads to FASTQ.<br>
            When FASTQ files are being used Albacore will not be used.
        </p>
        <h3>Assembly</h3>
        <p>
            Miniasm is used for assembly.<br>
            The file created by Miniasm containing the contigs will be used for further analysis.
        </p>
        <h3>Circularize contigs</h3>
        <p>
            Simple-Circularise is used to try and circularize the contigs when using Miniasm. <br>
            This option can be selected in the pipeline settings menu and after selecting the Miniasm assembler and is set to OFF by default. <br>
        </p>
        <h3>Plasmid Detection</h3>
        <p>
            BLAST+ is used for plasmid detection.<br>
            BLAST can be used with a remote database (NT or NR) or with a local NT or plasmid database.
        </p>
        <h3>Antibiotic Resistance</h3>
        <p>
            resFinder is used to find antibiotic resistance genes based on the contig file created by the assembler.<br>
            The user can select a specific gene to search or to find all available resistance genes present on the contig.<br>
            The minimum length cutoff can be entered when using resFinder, this option can be selected from the pipeline menu <br>
            and has the options 80%, 85%, 90%, 95% and 100%.<br>
            The % of identity cutoff can be selected from the pipeline menu, the option are 60%, 70%, 80%, 90% and 100%.
        </p>
        <h3>Incompatibility factors</h3>
        <p>
            plasmidFinder is used to find incompatibility factors based on the contig file created by the assembler.<br>
            The enterobacteriaceae database will be used by default.
        </p>
    </div>
    <div id="read_results">
        <h2>View Results</h2>
        <h3>QC</h3>
        <p>
            A QC report will be shown at the top of the result page. Click on the Show QC button to view the report.<br>
            The report contains a general summary with the mean read length, median read length, number of reads,<br>
            read length N50 and total bases. This report includes the following plots: 
            <ul>
                <li>Histogram of read length</li>
                <li>Histogram of read lengths after log transformation</li>
                <li>Weighted Histogram of read lengths</li>
                <li>Weighted Histogram of read lengths after log transformation</li>
                <li>Yield by length</li>
            </ul>
        </p>
        <h3>Antibiotic Resistance, incompatibility factors and BLAST results</h3>
        <p>
            The results will be presented in different headers (Assembly, Resfinder, BLAST and Plasmids). <br>
            The Assembly header will show the contigs created by Miniasm and the length of these contigs. <br>
            The Resfinder header will show the antibiotic resistance genese found per contig.<br>
            It will also show the percentage identity of the found gene.<br>
        </p>
    </div>
    <div style="clear:both;"></div>
</div>
{% endblock %}